rgo/component.py

```diff
@@ -95,94 +93,31 @@ def make_srpm(self, tmpdir):
             spec_name = self.git.spec_path or "{!s}.spec".format(self.name)
             patches = PatchesAction.keep
 
-        # extract spec file from specified branch and save it in temporary location
-        spec = os.path.join(workdir, "original.spec")
-        with open(spec, "w") as f_spec:
-            subprocess.run(["git", "cat-file", "-p", "{}:{}".format(spec_git.ref, spec_name)],
-                           cwd=spec_git.cwd, check=True, stdout=f_spec)
-
-        rpmspec = rpm.spec(spec)
-
-        spec_name = rpmspec.sourceHeader["Name"]
-        if isinstance(spec_name, bytes):
-            spec_name = spec_name.decode("utf-8")
+        original_spec_path = utils.extract_specfile_from_branch(workdir, spec_git, spec_name)
+        final_spec_path = utils.prepare_spec_and_archive(workdir, original_spec_path, self.git,
+                                                         self.name, self.version_from, patches)
+        if patches == PatchesAction.keep:
+            utils.prepare_specifle_patches(workdir, final_spec_path, self.distgit)
+        srpm_path = utils.rpmbuild_srpm(workdir, final_spec_path)
 
-        spec_version = rpmspec.sourceHeader["Version"]
-        if isinstance(spec_version, bytes):
-            spec_version = spec_version.decode("utf-8")
-
-        _spec_path = os.path.join(workdir, "{!s}.spec".format(spec_name))
+        self.srpm = os.path.join(tmpdir, os.path.basename(srpm_path))
+        shutil.move(srpm_path, self.srpm)
+        LOGGER.info("Built (no)source RPM: %r from %r", self.srpm, self.git.ref_info())
 
-        if self.git:
-            version, release = self.git.describe(self.name, spec_version=spec_version, version_from=self.version_from)
+        # Build overrides srpms
+        for override_git in self.overrides:
+            spec_name = override_git.spec_path or "{!s}.spec".format(self.name)
+            original_spec_path = utils.extract_specfile_from_branch(workdir, override_git, spec_name)
+            final_spec_path = utils.prepare_spec_and_archive(workdir, original_spec_path, self.git, self.name,
+                                                             self.version_from, override_git.patches)
+            if override_git.patches == PatchesAction.keep:
+                utils.prepare_specifle_patches(workdir, final_spec_path, override_git)
+            srpm_path = utils.rpmbuild_srpm(workdir, final_spec_path)
```

Comment:
I like how you moved the code from here to the various functions, but ultimately there's still this bit that's repeated inside the `for` loop and also called on the main srpm before it. So this seems like this should go into a function (a private method of this class, preferably?), and then it's possibly not even necessary to maintain the other functions you've created, since they'll be only called from a single place. The full business logic could just be in that single method (it'll be a bit busy, but I personally prefer that to another level of indirection).

Reply:
I have started with that but then changed it to this, I like how the steps are named by the functions. But I don't mind changing it back and adding comments instead.

Reply:
If you want to keep it split in multiple functions, fine with me. But do you also dislike to have a single function for that duplicate piece of code above the `for` loop and inside it?

Reply:
In the end I put it into a single bigger function with comments that also de-duplicates the code.